language/move-vm/runtime/src/runtime.rs

```diff
@@ -177,24 +264,27 @@ impl VMRuntime {
         module: &ModuleId,
         function_name: &IdentStr,
         ty_args: Vec<TypeTag>,
-        args: Vec<Value>,
+        args: Vec<Vec<u8>>,
+        senders: Vec<AccountAddress>,
         data_store: &mut impl DataStore,
         cost_strategy: &mut CostStrategy,
         log_context: &impl LogContext,
     ) -> VMResult<()> {
         // load the function in the given module, perform verification of the module and
         // its dependencies if the module was not loaded
-        let (func, type_params) =
+        let (func, ty_args, params) =
             self.loader
                 .load_function(function_name, module, &ty_args, data_store, log_context)?;
 
         // check the arguments provided are of restricted types
-        check_args(&args).map_err(|e| e.finish(Location::Module(module.clone())))?;
+        let args = self
+            .create_signers_and_arguments(&params, senders, args)
```

Comment:
What will happen if I try to call a function with a signature like `f(i: u64, s: &signer)`? Although we force the `signer`s to occur first in scripts, functions do not have this restriction.

Reply:
Good question. In order to address this I added a few new error codes and slightly changed the way the checks were performed.

Here's how we do it now:
- First we look at the signature of the function and split the params into two parts: `[prefix signer refs] [rest]`
- If `len([prefix signer refs])` is non-zero and not equal to `len([signer args])` then `Err(NUMBER_OF_SIGNER_ARGUMENTS_MISMATCH)`
- if `len(args)` != `len([rest])` then `Err(NUMBER_OF_ARGUMENTS_MISMATCH)`
- We try to obtain the layouts of the types in `[rest]`, and if one fails, we return `Err(INVALID_PARAM_TYPE_FOR_DESERIALIZATION)`. This error code covers not only `&signer` but also other reference or even invalid types.
- Lastly we try to deserialize the values and return `Err(FAILED_TO_DESERIALIZE_ARGUMENT)` upon failure.

BTW I fixed a few existing tests but that was just to see how it runs. Still planning to review the tests again and will likely rewrite a few of them and add more. Will let you know when that is done.

Reply:
I see. That seems a bit funky, because a signer can be both in `senders` and `args`...

I wonder if decoupling the script/function cases and checking signature validity from deserialization would make life easier. For example:

- Replace `create_signers_and_arguments(params, senders, args)` with `senders.extend(args); let all_args = senders; create_arguments(senders, args)`. All `create_arguments` does is match types to args + serialize, and it treats `&signer` and `address` the same.
- Add a separate function for checking that the `signer`s are the first part of the signature + call it only from the `script` case.